mama/c_cpp/src/c/dqstrategyplugin.c

```diff
@@ -0,0 +1,920 @@
+
+        strategyImpl->mTryToFillGap = 1;
+
+        msgStatus = mamaMsgStatus_statusForMsg (msg);
+        /* Check the status of the message.  If it is stale,
+           do not request a recap and do not set status OK. */
+        if ((msgStatus == MAMA_MSG_STATUS_POSSIBLY_STALE) ||
+            (msgStatus == MAMA_MSG_STATUS_STALE))
+        {
+            ctx->mDQState = DQ_STATE_STALE_NO_RECAP;
+            resetDqContext (ctx, seqNum, senderId);
+            return MAMA_STATUS_OK;
+        }
+    case MAMA_MSG_TYPE_RECAP        :
+    case MAMA_MSG_TYPE_BOOK_RECAP   :
+        /* For late joins or middlewares that support a publish cache, it is possible that you will get old updates
+           in this case take no action */
+        if (DQ_SCHEME_INGORE_DUPS == mamaTransportImpl_getDqStrategyScheme(transport))
+        {
+            if (MAMA_MSG_TYPE_RECAP == msgType)
+            {
+                /* Feed-handlers maintain sequence number for a Record FT Recap. */
+                if ((seqNum <= ctxSeqNum) && ((ctx->mDQState != DQ_STATE_WAITING_FOR_RECAP) &&
+                                              (ctx->mDQState != DQ_STATE_WAITING_FOR_RECAP_AFTER_FT)))
+                {
+                    ctx->mDoNotForward = 1;
+                    return MAMA_STATUS_OK;
+                }
+            }
+            else if (MAMA_MSG_TYPE_BOOK_RECAP == msgType)
+            {
+                if (0 == seqNum && ctxSeqNum > 0)
+                {
+                    /* Special case of an FT Order Book Recap where a SeqNum of 0 is used. */
+                    if (ctx->mDQState != DQ_STATE_WAITING_FOR_RECAP_AFTER_FT)
+                    {
+                        ctx->mDoNotForward = 1;
+                        return MAMA_STATUS_OK;
+                    }
+                }
+                /* Solicited Recap from Feed-Handler or
+                 * solicited / unsolicited Recap from mid-tier. */
+                else if ((seqNum <= ctxSeqNum) && ((ctx->mDQState != DQ_STATE_WAITING_FOR_RECAP) &&
+                                                   (ctx->mDQState != DQ_STATE_WAITING_FOR_RECAP_AFTER_FT)))
+                {
+                    ctx->mDoNotForward = 1;
+                    return MAMA_STATUS_OK;
+                }
+            }
+        }
+
+        if (mamaTransportImpl_preRecapCacheEnabled (transport))
+        {
+            strategyImpl->mTryToFillGap = 1;
+        }
+        mamaSubscription_unsetAllPossiblyStale (subscription);
+        resetDqState (strategy, ctx);
+        resetDqContext (ctx, seqNum, senderId);
+        ctx->mDoNotForward = 0;
+        return MAMA_STATUS_OK;
+    case MAMA_MSG_TYPE_DDICT_SNAPSHOT : /*No DQ checking for Datadictionary*/
+        return MAMA_STATUS_OK;
+    }
+
+    return MAMA_STATUS_OK;
+}
+
+mama_status
+dqstrategyMamaPlugin_shutdownHook (mamaPluginInfo pluginInfo)
+{
+    free(pluginInfo);
+    mama_status status = MAMA_STATUS_OK;
+
+    return status;
```

Comment:
Minor point but should just be "return MAMA_STATUS_OK" instead of being over two lines